hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClientThrottlingInterceptFactory.java

```diff
@@ -0,0 +1,59 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.fs.azurebfs.services;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+final class AbfsClientThrottlingInterceptFactory {
+
+  private AbfsClientThrottlingInterceptFactory() {
+  }
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+      AbfsClientThrottlingInterceptFactory.class);
+
+  private static Map<String, AbfsClientThrottlingIntercept> instanceMapping
+      = new ConcurrentHashMap<>();
+
+  static synchronized AbfsClientThrottlingIntercept getInstance(String accountName,
+      boolean isAutoThrottlingEnabled,
+      boolean isSingletonEnabled) {
+    AbfsClientThrottlingIntercept instance;
+    if (isSingletonEnabled) {
+      instance = AbfsClientThrottlingIntercept.initializeSingleton(
+          isAutoThrottlingEnabled);
+    } else {
+      if (!isAutoThrottlingEnabled) {
+        return null;
+      }
+      if (instanceMapping.get(accountName) == null) {
+        LOG.debug("The accountName is: {} ", accountName);
+        instance = new AbfsClientThrottlingIntercept(accountName);
+        instanceMapping.put(accountName, instance);
```

Comment:
do a get() afterwards, in case a separate thread registers the interceptor while this code is creating its instance

Reply:
The method getInstance is a synchronized method and the instanceMapping map is also a concurrent hash map. We are also checking for null in line no 49, so race condition is ideally not possible here.